### DISCUSSION TITLE
C# Properties And Fields

### DISCUSSION THREAD

**Discussion by mishelshaji** [mishelshaji/Learn-DotNet#19] [Q&A] [answered]
Explain:

- What is a property?
- How to define a property?
- What is a field?
- How to define a field?
- What is the difference between a property and a field?
- What is a getter and setter?
- What benefits do properties have over fields?
- How to write getters and setters?

**Comment by sarath534** [comment]
# Fields 
Fields are normal variable members of a class. Generally, you should declare your fields as private, then use Properties to get and set their values. By this way you won’t affect their values them directly. This is common case practice since having public members violates the Encapsulation concept in OOP.
```csharp
public class Student
   {
       private int _id;  //fields
       private string _name;
   }
```
# Properties
They are actually special methods called “accessors”. Properties are called accessors because they offer a way to get and set a field if you have a private field. They have two codes inside; set{}; and get{}; called property accessors.
```csharp
public class Student
    {
        private int _id;
        private string _name;
        public int Id { get { return _id; } set { _id = value; } } // properties
        public string Name { get { return _name; } set { _name = value; } }
    }
```
### Difference between fields and properties:
The key difference between field and property in C# is that a field is a variable of any type that is declared directly in the class while property is a member that provides a flexible mechanism to read, write or compute the value of a private fields

### Getter and Setter:
 We use to access and change information stored in private fields. As their names imply, getters "get" information and setters "set" information.

**Comment by thanoofnaina** [comment, marked as answer]
### Property:
A property is a member that provides a flexible mechanism to read, write, or compute the value of a private field. Properties can be used as if they are public data member, but they are special methods called accessors. This feature enables data to be accessed easily and still helps promote the safety and flexibility of methods.

```csharp
public class mobile
{
   private string _imei;
   //Property.
   public string Imei { get { return _imei } set { _imei = value; }}
}
```

### Field:
A field is a variable of any type that is declared directly in a class. Fields are members of their containing type.

```csharp
public class mobile
{
   // Private field.
   private DateTime _dateofsale;
   // Public property exposes _date field safely.
   public DateTime DateOfSale
   {
      get { return _date }
      set { _date = value; }
   }
}
```

### Difference between property and field:

- A field is a variable of any type that is declared directly in a class. 
  A property is a member that provides a flexible mechanism to read, write or compute the value 
  of a private field.
- A field can be used to explain the characteristics of an object or a class.
  A property can be used to set and receive values of a field.
 
### Getter and Setter:

Simply getters and setters can be defined as:
Fields can be made read-only by using only getter.
Fields can be made write-only by using only setter.

Getters give public access to private data, They can also make a small modification of the returned result.
Setters allow for a private variable to be modified. They are important since they can provide validation before a value is set.

### Benefits of properties over fields:

- The setters and getters can be modified making breaking API changes, this is the primary reason for using properties over fields.
- Properties gives a level of abstraction to change the fields while not affecting how a class uses 
   them.
- Before allowing a change in data, the properties can validate the data.

### Writing getters and setters:

```csharp
public class mobile
{
   private string _brand;
   public string Brand { get { return _brand;} set { _brand = value} }
}
```

**Comment by jeweljose22** [comment]
## Property
A property is a member of a class that provides a flexible mechanism to read, write, or compute the value of a private field.
Properties are typically used as a public interface to an object's internal state.
In C# and other languages that support the concept of properties, they can be defined using the property keyword, followed by the type and the name of the property.
## Field
A field is a variable that is a member of a class or struct.
A field can be defined using the field keyword and the data type, followed by the name of the field.
## Difference 
The main difference between a property and a field is that properties have accessors (getters and setters) which are used to get and set the value of the private field, while fields can be directly accessed and modified.
## Getter & Setter
A getter is a method that is used to retrieve the value of a property, while a setter is a method that is used to set the value of a property.

## Benefits 

They provide a level of abstraction, allowing the internal representation of the data to change without affecting the external interface.
They can include additional logic, such as validation, in the getter and setter methods.
They can be read-only or write-only, while fields are always read-write.
To write a getter and setter, you can define a property with a get; and set; accessors, respectively.

```csharp
class MyClass {
    private int _myField;
    public int MyProperty {
        get { return _myField; }
        set { _myField = value; }
    }
}
```
The property in this example is MyProperty, and  _myField is the private field . The setter changes the value of _myField while the getter returns its value.

**Comment by Jiby-Joseph-123** [comment]
# Property

The property does not have storage locations. The property has accessors that contain the executable statements to read the values and to set the values. The accessor declaration can contain a get accessor and set accessor.

Assume that there is a class name Employee and it contains the properties such as EmployeeNo, Name, and Department.  We can use properties to set and get values.

Example:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EmployeeProject
{
    internal class Employee
    {
	// properties

        public int EmployeeNo { get; set; }  
        public string Name { get; set; }
        public string Department { get; set; }

        public Employee()
        {
            EmployeeNo= 0;
            Name = "Anna";
            Department = "QA";
        }

        public override string ToString()
        {
            return $"Name: {Name}, Employee Number: {EmployeeNo}, Department: {Department}";
        }

    }
  
    }

using EmployeeProject;

var employee = new Employee();
Console.WriteLine(employee);

```

# Field

The characteristics of the object are described by the fields. An Employee object can have fields such as employee number, name, and department. A field is a variable of any type that is declared directly in a class.

Example:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EmployeeProject
{
    internal class Employee
    {
        private int _employeeNo;
        public string Name;
        public string Department;
  

        public override string ToString()
        {
            return $"Name: {Name}, Employee Number: {_employeeNo}, Department: {Department}";
        }
    }
  
    }

using EmployeeProject;

var employee = new Employee();
employee.Name = "Test";
employee.Department = "QA";
Console.WriteLine(employee);
```
# Differnce between Property and Field

Fields and properties are two terms used in C# OOP. The difference between a field and a property is that a field is a variable of any type that is declared directly in the class while a property is a member that provides a flexible mechanism to read, write or compute the value of a private field.
A field can be used to explain the characteristics of an object or a class. A property can be used to set and receive values of a field.

**Comment by Adarshv12** [comment]
# Properties

Property in C# is a member of a class that provides a flexible mechanism for classes to expose private fields. Internally, C# properties are special methods called accessors. A C# property has two accessors, get the property accessor and set the property accessor. A get accessor returns a property value, and a set accessor assigns a new value. The value keyword represents the value of a property.

```csharp
class MyClass  
{  
    private int x;  
    public void SetX(int i)  
    {  
        x = i;  
    }  
    public int GetX()  
    {  
        return x;  
    }  
}  
class MyClient  
{  
    public static void Main()  
    {  
        MyClass mc = new MyClass();  
        mc.SetX(10);  
        int xVal = mc.GetX();  
        Console.WriteLine(xVal);
    }  
}
```

# Fields

A field, in C#, is a member of a class or an object of any type that represents a memory location for storing a value.

Fields are used to store data that must be accessible to multiple methods of a class and available throughout the lifetime of an object. Fields enable a class or struct to encapsulate the data with options to specify its accessibility at multiple levels.

```csharp
public class Fields1
{
	private string name = "John Doe";
	private int age = 42;

	public void Describe()
	{
		string description = this.name + " is " + this.age + " years old...";
		Console.WriteLine(description);
	}
}

```
**Difference**

In OOP, the program or the software can be modelled using objects. The objects are created using classes. A class is a blueprint to create objects. Fields and properties are two terms used in C# OOP. This article discussed the difference between field and property in C#. The difference between field and property in C# is that a field is a variable of any type that is declared directly in the class while property is a member that provides a flexible mechanism to read, write or compute the value of a private field.

# Getter and Setter

Getters and setters are methods used to declare or obtain the values of variables, usually private ones. They are important because it allows for a central location that is able to handle data prior to declaring it or returning it to the developer. Within a getter or setter you are able to consistently handle data that will eventually be passed into a variable or additional functions.

### Benefit of properties over fields

You can delay the creation of actual reference fields until you use them, which saves resources. You can differentiate the representation and actual storage. Representation is implemented via properties and storage is implemented via fields. You can check constraints when setting and getting properties. If the value is not suitable, you do not store the data in the field and a type-safety error is returned. This really provides 100% type-safe accessors on demand.

#### Example of setter and getter

```csharp
class Person
{
  private string name; // field
  public string Name   // property
  {
    get { return name; }
    set { name = value; }
  }
}

class Program
{
  static void Main(string[] args)
  {
    Person myObj = new Person();
    myObj.Name = "Liam";
    Console.WriteLine(myObj.Name);
  }
}
```

**Comment by sreelakshmiChandran** [comment]
# Properties

Property is a member that provides a flexible mechanism to read, write, or compute the value of a private field.
Properties can be used as if they are public data members, but they are actually special methods called accessors.
The methods that implement a property are called get and set accessors.
```csharp
public student
{
   public string Name { get; set; }

   private string _age;
   public int Age { get {return _age} set{_age = value }}
}
```
- The get accessor returns the value of the private field.
- The set accessor assigns a new value to the private field.

# Field

A field is a variable that is associated with a class or struct. Fields are declared inside a class or struct but outside of any method.
```csharp
class Person
{
    public string Name;
    public int Age;

    public void Display()
    {
        Console.WriteLine(Name + "is " + Age + "years old");
    }
}
```
## Property Vs Field
- Access to the field is restricted and that of properties is not restricted.
- A field is used to explain the characteristics of an object or a class and property is used to set and receive values of a field.
- Field is directly declared in the class.

## Getter
Getter sets the field as read-only. The users can only read the field value. Getters give the public access to private data, They can also make a small modification to the returned result.

## Setter
The setter sets the field as write-only. The users can set the field value. Setters allow for a private variable to be modified. They are important since they can provide validation before a value is set.

```csharp
public student
{
   public string Name { get; set; }

   private string _age;
   public int Age { get {return _age} set{_age = value }}
}
```

**Comment by kavitha-BS** [comment]
**property**
     
      A property is a member that provides a flexible mechanism to read, write, or compute the value of a private field. Properties can be used as if they're public data members, but they're special methods called accessors. This feature enables data to be accessed easily and still helps promote the safety and flexibility of methods.
eg:
    ```  public class SaleItem
      {
    public string Name
    { get; set; }
   public decimal Price
    { get; set; }
  }  ```

**Field**
     A field is a variable of any type that is declared directly in a class or struct. Fields are members of their containing type.
A class or struct may have instance fields, static fields, or both. Instance fields are specific to an instance of a type.

eg:
``` 
public class Fields1
{
	private string name = "John Doe";
	private int age = 42;

	public void Describe()
	{
		string description = this.name + " is " + this.age + " years old...";
		Console.WriteLine(description);
	}
} 
```
**Difference between property and fields**

- A field is a variable of any type that is declared directly in a class.

- A property is a member that provides a flexible mechanism to read, write or compute the value of a private field.

- A field can be used to explain the characteristics of an object or a class.

- A property can be used to set and receive values of a field.

**Benefits of property over field**

-     It allows for versioning if later you need extra logic. Adding logic to the getter or setter won't break existing code.

-     It allows data binding to work properly (most data binding frameworks don't work with fields).

**Getter and Setter**

The Setter method is used to set or assign values to a property(example:- name) of a class. And the Getter method is used to get the values of a property(example:- name) in a class.
eg:
```
class Human{

	string name;
	string food;
	string language;
	
	public string Name 
	{
		get { return name; }
		set { name = value; }
	}
	
	public string Food 
	{
		get { return food; }
		set { food = value; }
	}
	
	public string Language 
	{
		get { return language; }
		set { language = value; }
	}

	public void eat(){
		System.Console.WriteLine("Eats "+food);
	}

	public void speak(){
		System.Console.WriteLine("Speaks "+language);
	}
}
```
